File: src/Billing.Config/TenantConfigSync/Model/Tenant.cs
namespace Billing.Config.TenantConfigSync.Model;

/// <summary>
/// A node in the multi-tenant tree — the .NET mirror of routesphere's <c>Tenant</c>.
/// The immutable data (name, dbName, parent, children, context) comes from config-manager;
/// the computed lookups (<see cref="Index"/>, <see cref="AncestorChain"/>) are filled once,
/// after the tree is loaded, by the tree builder in this package's Internal layer.
///
/// <para><b>dbName</b> is the globally-unique key (admin = the operator DB, resellers = res_NNN);
/// idPartner is unique only within a DB, so the chain is keyed by dbName.</para>
/// </summary>
public sealed class Tenant
{
    public required string Name { get; init; }
    public required string DbName { get; init; }

    /// <summary>Parent dbName; null = root (admin/operator).</summary>
    public string? Parent { get; init; }

    public IReadOnlyDictionary<string, Tenant> Children { get; init; }
        = new Dictionary<string, Tenant>();

    /// <summary>This tenant's config snapshot — holds the MediationContext.</summary>
    public DynamicContext Context { get; init; } = DynamicContext.Empty;

    // --- computed after load (one shared index across the whole tree; O(1) dbName lookup) ---

    /// <summary>dbName → Tenant, shared across the whole tree. Set by the tree builder.</summary>
    public IReadOnlyDictionary<string, Tenant> Index { get; internal set; }
        = new Dictionary<string, Tenant>();

    /// <summary>[this, parent, …, root]. Set by the tree builder; the reseller charge chain.</summary>
    public IReadOnlyList<Tenant> AncestorChain { get; internal set; } = [];
}

File: java/src/main/java/com/telcobright/billing/tenantconfigsync/model/Tenant.java
package com.telcobright.billing.tenantconfigsync.model;

import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * A node in the multi-tenant tree — the .NET mirror of routesphere's {@code Tenant}.
 * The immutable data (name, dbName, parent, children, context) comes from config-manager;
 * the computed lookups ({@link #Index}, {@link #AncestorChain}) are filled once,
 * after the tree is loaded, by the tree builder in this package's Internal layer.
 *
 * <p><b>dbName</b> is the globally-unique key (admin = the operator DB, resellers = res_NNN);
 * idPartner is unique only within a DB, so the chain is keyed by dbName.
 *
 * <p>Faithful-port note: C# {@code required}/{@code init}/{@code internal set} properties become
 * public mutable fields (the mapper sets Name/DbName/Parent/Children/Context; the tree builder
 * later writes Index/AncestorChain). Field names kept verbatim.
 */
public final class Tenant {
    public String Name;
    public String DbName;

    /** Parent dbName; null = root (admin/operator). */
    public String Parent;

    public Map<String, Tenant> Children = new HashMap<>();

    /** This tenant's config snapshot — holds the MediationContext. */
    public DynamicContext Context = DynamicContext.Empty;

    // --- computed after load (one shared index across the whole tree; O(1) dbName lookup) ---

    /** dbName → Tenant, shared across the whole tree. Set by the tree builder. */
    public Map<String, Tenant> Index = new HashMap<>();

    /** [this, parent, …, root]. Set by the tree builder; the reseller charge chain. */
    public List<Tenant> AncestorChain = List.of();
}
